src/chain/auth.sw

```diff
@@ -1,28 +1,39 @@
 library auth;
+//! Functionality for determining who is calling an ABI method
 
-// this can be a generic option when options land
-enum Caller {
-    Some: b256,
-    None: (),
+use ::result::Result;
+use ::address::Address;
+use ::contract_id::ContractId;
+
+pub enum AuthError {
+    ContextError: (),
+}
+
+pub enum Sender {
+    Address: Address,
+    Id: ContractId,
 }
 
-/// Returns `true` if the caller is external.
+/// Returns `true` if the caller is external (ie: a script or predicate).
 pub fn caller_is_external() -> bool {
     asm(r1) {
         gm r1 i1;
         r1: bool
     }
 }
 
-pub fn caller() -> Caller {
-    // if parent is not external
+/// Returns a Result::Ok(Address) or Result::Error.
+// NOTE: Currently only returns the Ok variant of result if the parent context is Internal.
+pub fn msg_sender() -> Result<Sender, AuthError> {
     if !caller_is_external() {
-        // get the caller
-        Caller::Some(asm(r1) {
-            gmr1i2;
+        // Get caller's contract ID
+        let id = ~ContractId::from(asm(r1) {
+            gm r1 i2;
```

Comment:
Can you make constants for each `GM` immediate, then pass in the constant into the asm block?